testing/_diff.ts

```diff
@@ -226,3 +226,80 @@ export function diff<T>(A: T[], B: T[]): Array<DiffResult<T>> {
     ),
   ];
 }
+
+/**
+ * Renders the differences between the actual and expected strings
+ * Partially inspired from https://github.com/kpdecker/jsdiff
+ * @param A Actual string
+ * @param B Expected string
+ */
+export function diffstr(A: string, B: string) {
+  function tokenize(string: string, { wordDiff = false } = {}): string[] {
+    if (wordDiff) {
+      // Split string on whitespace symbols
+      const tokens = string.split(/([^\S\r\n]+|[()[\]{}'"\r\n]|\b)/);
+      // Extended Latin character set
+      const words =
+        /^[a-zA-Z\u{C0}-\u{FF}\u{D8}-\u{F6}\u{F8}-\u{2C6}\u{2C8}-\u{2D7}\u{2DE}-\u{2FF}\u{1E00}-\u{1EFF}]+$/u;
+
+      // Join boundary splits that we do not consider to be boundaries and merge empty strings surrounded by word chars
+      for (let i = 0; i < tokens.length - 1; i++) {
+        if (
+          !tokens[i + 1] && tokens[i + 2] && words.test(tokens[i]) &&
+          words.test(tokens[i + 2])
+        ) {
+          tokens[i] += tokens[i + 2];
+          tokens.splice(i + 1, 2);
+          i--;
+        }
+      }
+      return tokens.filter((token) => token);
+    } else {
+      // Split string on new lines symbols
+      const tokens = [], lines = string.split(/(\n|\r\n)/);
+
+      // Ignore final empty token when text ends with a newline
+      if (!lines[lines.length - 1]) {
+        lines.pop();
+      }
+
+      // Merge the content and line separators into single tokens
+      for (let i = 0; i < lines.length; i++) {
+        if (i % 2) {
+          tokens[tokens.length - 1] += lines[i];
+        } else {
+          tokens.push(lines[i]);
+        }
+      }
+      return tokens;
+    }
+  }
+
+  // Try to autodetect the most suitable string rendering, starting by words-diff
+  let wordDiff = true;
+  let diffResult = diff(tokenize(A, { wordDiff }), tokenize(B, { wordDiff }));
+  const common =
+    diffResult.filter(({ type, value }) =>
+      type === DiffType.common && value.trim()
+    ).length;
+  const edited =
+    diffResult.filter(({ type }) =>
+      type === DiffType.added || type === DiffType.removed
+    ).length;
+
+  // If edited ratio is too high switch to multi-line diff instead
+  // We add a trailing newline to ensure that last tokens are displayed correctly
+  if (
+    edited / (common + edited) >
+      0.85 **
+        Math.max(1, A.match(/\n/g)?.length ?? 0, B.match(/\n/g)?.length ?? 0)
```

Comment:
Not sure whether it's the best way to make this but here's the reasoning:

- `edited / (common + edited)` give a ratio of "changes performed"
- `Math.max(1, A.match(/\n/g)?.length ?? 0, B.match(/\n/g)?.length ?? 0)` tells how many lines input has
- `0.85 ** x` (where `x` is above value) is the threshold from switching to word-diff to multiline-diff
  
Thus the more lines you have, the more likely you'll end up with multiline-diff unless there are only a few changes (like a word or two) whereas when there are less changes present the word-diff will be prefered